tests/integration/integration_test.go

```diff
@@ -50,6 +51,23 @@ func TestEmptyGo(t *testing.T) {
 	})
 }
 
+// Tests emitting many engine events doesn't result in a performance problem.
+func TestEngineEventStress(t *testing.T) {
+	start := time.Now()
+	integration.ProgramTest(t, &integration.ProgramTestOptions{
+		Dir:          "ee_stress",
+		Dependencies: []string{"@pulumi/pulumi"},
+		Quick:        true,
+		ExtraRuntimeValidation: func(t *testing.T, stackInfo integration.RuntimeValidationStackInfo) {
+			// Prior to pulumi/pulumi#2303, a preview or update would take ~40s.
+			// Since then, it should now be down to ~4s, with additional padding.
+			if time.Since(start) > 8*time.Second {
+				t.Errorf("Performance regression in test ee_stress. Took %s to execute.", time.Since(start))
+			}
+		},
+	})
+}
```

Comment:
Instead of manually timing things here - which will include all of the dozen or so steps the intergation framework runs (including things link `yarn install`) - you could use the `ReportStats:` field on the ProgramTestOptions to specify a callback that will be called with timing data for every one of the individual commands that were run during the test run.  You can then make more fine-grained assertions about each individual step (preview and update independently even).